### ISSUE TITLE
Station mode broken with sdk v1.0.1_b2

### ISSUE THREAD

**Issue by empatzero** [kadamski/esp-lwip#7]
The new espressif sdk v1.0.1_b2 is causing troubles.

In STATIONAP_MODE i get this messages, with both USE_OUR_LWIP_IF options:

    dhcp server start:(ip:192.168.3.1,mask:255.255.255.0,gw:192.168.3.1)
    add if1
    bcn 100
    scandone
    reconnect
    scandone
    reconnect
    scandone
    ...

and the station does not connect to AP.

**Comment by empatzero** [comment]
Same with esp_iot_sdk_v1.0.1_15_04_24.

**Comment by igrr** [comment]
I've seen the same behavior with Espressif's version of lwip, since SDK 0.9.3 or even earlier. Never figured out what was the cause. It has only occured with some combinations of a router and a module.

**Comment by anakod** [comment]
>It has only occured with some combinations of a router and a module.
No, unfortunately STA isn't work on any routers since SDK v1.0.1_b2
Replacing library to Espressif binary solves this problem.

**Comment by kadamski** [comment]
I was busy lately and had no time for esp8266 but I'm finally able to look into this. Unfortunately does not look like a trivial issue.

**Comment by kadamski** [comment]
It looks like it was broken even in 1.0.1_b1. It looks like a lot has changed in that version and it would be a lot of work to reverse engineer all of this. Looks like `netif` structure has some new member and this may be direct reason why this isn't working. Seems like I wont be able to fix this so this lwip library will only be comatible with SDK up to 1.0.0 :/

**Comment by anakod** [comment]
>  Seems like I wont be able to fix this so this lwip library will only be comatible with SDK up to 1.0.0 :/

Really very bad news :(


May be you will look when you will have more time? Switching to your netif version also wouldn't help?


On our forum pvvx working on SDK decompilation. It should be related (but doesn't completed unfortunately):
```C++
    queue = (void *) pvPortMalloc(sizeof(struct ETSEventTag) * QUEUE_LEN); // pvPortZalloc(80)

    if (state->dhcps_if == 0) {
#if LWIP_NETIF_HOSTNAME
#ifdef LWIP_NETIF_HOSTNAME_PREFIX
        myif->hostname = os_malloc(sizeof(LWIP_NETIF_HOSTNAME_PREFIX)+10);
        os_sprintf(myif->hostname, "%s%02x%02x%02x", LWIP_NETIF_HOSTNAME_PREFIX,
                myif->hwaddr[3], myif->hwaddr[4], myif->hwaddr[5]);
#endif
#endif
        ??? = queue; // [0x3FFE9B84] = queue; // Some unknown address
        ets_task(task_if0, TASK_IF0_PRIO, (ETSEvent *)queue, QUEUE_LEN);
        netif_add(myif, &ipaddr, &netmask, &gw, state, init_fn, ethernet_input);
    } else {
    	???+4 = queue; //[0x3FFE9B88] = queue; // Some unknown address
        netif_set_addr(myif, &ipaddr, &netmask, &gw);

```

**Comment by kadamski** [comment]
I tried that change and it didn't resolve the issue. The problem is - even if I will find the issue new, Espressif can easily change something braking binary compatibility again. And I wont be able to keep up. They are also adding some more and more functionality to their lwip library which I'm also have no resources to follow.

**Comment by anakod** [comment]
We have a plan to completely reverse Espressif SDK.

**Comment by stv0g** [comment]
> We have a plan to completely reverse Espressif SDK.

@anakod That sound interesting. Do you've some more infos / contacts about your efforts?

@kadamski I've noticed that the RTOS SDK uses LWIP 1.4.1 and also includes the header files.
It seems to be a little more recent version of their LWIP fork than the one we have from ther IoT example.

Did you made a diff against those headers to find some diffrences?

**Comment by anakod** [comment]
@kadamski Did you look pvvx update:
http://www.esp8266.com/viewtopic.php?f=9&t=1492&start=30#p18680
https://github.com/pvvx/esp8266web/blob/master/app/main/eagle_lwip_if.c

It should completely solve new SDK compatibility issues.

**Comment by anakod** [comment]
> That sound interesting. Do you've some more infos / contacts about your efforts?
@stv0g main SDK reversing contact - @pvvx.
